index.html

```diff
@@ -101,19 +199,623 @@ <h3>
+            const { writable } = await socket.connection;
+            const writer = writable.getWriter();
+
+            writer.write(encoder.encode("A long message that will take..."));
+            await writer.close();
+            await socket.close();
+          </pre>
+
+          To discard any unsent data the application could instead call
+          `writer.`{{WritableStreamDefaultWriter/abort()}}.
+
+          If a loop is being used to read data from the socket, then it must be exited before calling `socket`.{{TCPSocket/close()}}.
+
+          <pre class="js">
+            const { readable } = await socket.connection;
+            const reader = readable.getReader();
+
+            async function readUntilClosed() {
+              while (true) {
+                const { value, done } = await reader.read();
+                if (done) {
+                  // |reader| has been canceled.
+                  break;
+                }
+                // Do something with |value|...
+              }
+
+              await socket.close();
+            }
+
+            const read_complete = readUntilClosed();
+
+            // Sometime later...
+            reader.cancel();
+            await read_complete;
+          </pre>
+
+        </aside>
+
+        The {{TCPSocket/close()}} method steps are:
+
+        <ol>
+          <li>
+            If [=this=].{{TCPSocket/[[connectionPromise]]}} is rejected or not yet resolved, [=reject=] with
+            "{{InvalidStateError}}" {{DOMException}}.
+          <li>
+            If [=this=].{{TCPSocket/[[closedPromise]]}} is already resolved, return [=this=].{{TCPSocket/[[closedPromise]]}}.
+          <li>
+            If [=this=].{{TCPSocket/[[readable]]}} or [=this=].{{TCPSocket/[[writable]]}} are <a href="https://streams.spec.whatwg.org/#lock">locked</a>,
+            [=reject=] with "{{InvalidStateError}}" {{DOMException}}.
+          <li>
+            Let |cancelPromise:Promise| be the result of invoking
+            [=ReadableStream/cancel=] on [=this=].{{TCPSocket/[[readable]]}}.
+          <li>
+            Mark |cancelPromise| as handled.
+          <li>
+            Let |abortPromise:Promise| be the result of invoking
+            [=WritableStream/abort=] on [=this=].{{TCPSocket/[[writable]]}}.
+          <li>
+            Mark |abortPromise| as handled.
+          <li>
+            [=Resolve=] [=this=].{{TCPSocket/[[closedPromise]]}} with `undefined`.
+          <li>Return [=this=].{{TCPSocket/[[closedPromise]]}}.
+        </ol>
+
+      </section>
+    </section>
+
+    <section data-dfn-for="UDPSocket">
+      <h2>
+        {{UDPSocket}} interface
+      </h2>
+
+      <pre class="idl">
+        [Exposed=(Window,Worker), SecureContext]
+        interface UDPSocket {
+          constructor(DOMString remoteAddress, unsigned short remotePort,
+              optional UDPSocketOptions options = {});
+
+          readonly attribute Promise&lt;UDPSocketConnection> connection;
```

Comment:
It's weird to call this a "connection" since UDP doesn't really have connections. What if we called this attribute "opened" for both TCP and UDP sockets? Then it would also work for a `TCPServerSocket` and match well with the `closed` attribute.

Reply:
That sounds reasonable, although renaming the dictionary to `TCPSocketOpened` is also a bit weird :) Or should we retain the type name as `TCPSocketConnection` and go with `opened` for the attribute name?

Reply:
Dictionary type names aren't visible to script so it doesn't really matter. Maybe `TCPSocketState` just for something generic enough that it would also work for UDP and server sockets? Or `TCPSocketOpenInfo`.

Reply:
Fixed!